scanpy/external/pp/_hashsolo.py

```diff
@@ -0,0 +1,383 @@
+        in the transcriptome space, e.g. UMI counts, pct mito reads, etc.
+    number_of_noise_barcodes : int
+        number of barcodes to used to calculated noise distribution
+    Returns
+    -------
+    bayes_dict_results : dict
+        "most_likely_hypothesis" key is a 1d np.array of the most likely hypothesis
+        "probs_hypotheses" key is a 2d np.array probability of each hypothesis
+        "log_likelihoods_for_each_hypothesis" key is a 2d np.array log likelihood of each hypothesis
+    """
+    priors = np.array(priors)
+    log_likelihoods_for_each_hypothesis, _, _ = _calculate_log_likelihoods(
+        data, number_of_noise_barcodes
+    )
+    probs_hypotheses = (
+        np.exp(log_likelihoods_for_each_hypothesis)
+        * priors
+        / np.sum(
+            np.multiply(np.exp(log_likelihoods_for_each_hypothesis), priors), axis=1
+        )[:, None]
+    )
+    most_likely_hypothesis = np.argmax(probs_hypotheses, axis=1)
+    return {
+        "most_likely_hypothesis": most_likely_hypothesis,
+        "probs_hypotheses": probs_hypotheses,
+        "log_likelihoods_for_each_hypothesis": log_likelihoods_for_each_hypothesis,
+    }
+
+
+def hashsolo(
+    cell_hashing_adata: anndata.AnnData,
+    cell_hashing_columns: list,
+    priors: list = [0.01, 0.8, 0.19],
+    pre_existing_clusters: str = None,
+    clustering_data: anndata.AnnData = None,
+    resolutions: list = [0.1, 0.25, 0.5, 0.75, 1],
+    number_of_noise_barcodes: int = None,
+    inplace: bool = True,
+):
+    """Probabilistic demultiplexing of cell hashing data using HashSolo [Bernstein20]_.
+
+    .. note::
+        More information and bug reports `here <https://github.com/calico/solo>`__.
+    Parameters
+    ----------
+    cell_hashing_adata : anndata.AnnData
+        Anndata object with cell hashes in .obs columns
+    cell_hashing_columns : list,
+        list specifying which columns in cell_hashing_adata.obs
+        are cell hashing counts
+    priors : list,
+        a list of your prior for each hypothesis
+        first element is your prior for the negative hypothesis
+        second element is your prior for the singlet hypothesis
+        third element is your prior for the doublet hypothesis
+        We use [0.01, 0.8, 0.19] by default because we assume the barcodes
+        in your cell hashing matrix are those cells which have passed QC
+        in the transcriptome space, e.g. UMI counts, pct mito reads, etc.
+    pre_existing_clusters : str
+        column in cell_hashing_adata.obs for how to break up demultiplexing
+        for example leiden or cell types, not batches though
+    inplace : bool
+        To do operation in place
+    Returns
+    -------
+    cell_hashing_adata : AnnData
+        if inplace is False returns AnnData with demultiplexing results
+        in .obs attribute otherwise does is in place
+
+    Examples
+    -------
+    >>> import anndata
```

Comment:
can you adapt the example to use `sce.pp.hashsolo` ?

Reply:
done

Reply:
I think this docstring example is missing the mandatory `cell_hashing_columns` arg?

Reply:
you are correct. @chris-rands